File: Sequence Valley/Heap/Q2. Find K Pairs with Smallest Sums/Q2_FindKPairswithSmallestSums.py
import heapq  # import heapq for min-heap

class Solution(object):
    def kSmallestPairs(self, nums1, nums2, k):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :type k: int
        :rtype: List[List[int]]
        """
        result = []  # list to store the final pairs
        if not nums1 or not nums2 or k == 0:  # edge case check
            return result

        # min-heap to store [sum, index in nums1, index in nums2]
        min_heap = []

        # Initialize heap with first element of nums2 paired with up to first k elements of nums1
        for i in range(min(len(nums1), k)):
            heapq.heappush(min_heap, (nums1[i] + nums2[0], i, 0))  # push sum and indexes

        # Extract k smallest pairs
        while k > 0 and min_heap:
            sum_val, i, j = heapq.heappop(min_heap)  # get smallest sum pair
            result.append([nums1[i], nums2[j]])      # add actual pair to result
            k -= 1

            # If there is a next element in nums2 for current nums1[i], push new pair
            if j + 1 < len(nums2):
                heapq.heappush(min_heap, (nums1[i] + nums2[j + 1], i, j + 1))

        return result  # return list of k smallest pairs

File: Sequence Valley/Heap/Q2. Find K Pairs with Smallest Sums/Q2_FindKPairswithSmallestSums.java
import java.util.*;  // import utility classes for List, ArrayList, PriorityQueue, Arrays

class Solution {
    public List<List<Integer>> kSmallestPairs(int[] nums1, int[] nums2, int k) {
        List<List<Integer>> result = new ArrayList<>();  // store final pairs
        if (nums1.length == 0 || nums2.length == 0 || k == 0) return result; // edge case check

        // Min-heap to store pairs with their sum
        // Each element = [sum of pair, index in nums1, index in nums2]
        PriorityQueue<int[]> minHeap = new PriorityQueue<>((a, b) -> a[0] - b[0]);

        // Initially, pair first element of nums2 with each element of nums1 (up to k elements)
        for (int i = 0; i < Math.min(nums1.length, k); i++) {
            minHeap.offer(new int[]{nums1[i] + nums2[0], i, 0}); // push sum and indexes into heap
        }

        // Extract k smallest pairs from the heap
        while (k > 0 && !minHeap.isEmpty()) {
            int[] top = minHeap.poll();  // remove the pair with the smallest sum
            int i = top[1], j = top[2];  // get indexes of nums1 and nums2
            result.add(Arrays.asList(nums1[i], nums2[j]));  // add the actual pair to result list
            k--;  // one pair has been used

            // If next element exists in nums2, push new pair with same nums1[i] and next nums2[j+1]
            if (j + 1 < nums2.length) {
                minHeap.offer(new int[]{nums1[i] + nums2[j + 1], i, j + 1});
            }
        }

        return result;  // return the final list of k smallest pairs
    }
}
